test/object-store/realm.cpp

```diff
@@ -618,12 +619,13 @@ TEST_CASE("Get Realm using Async Open", "[asyncOpen]") {
         ThreadSafeReference realm_ref;
         std::atomic<bool> called{false};
         auto task = Realm::get_synchronized_realm(config);
-        task->start([&](auto ref, auto error) {
+        realm::util::EventLoopDispatcher callback_handler([&](ThreadSafeReference&& ref, std::exception_ptr error) {
             std::lock_guard<std::mutex> lock(mutex);
             REQUIRE(!error);
             called = true;
             realm_ref = std::move(ref);
         });
+        task->start(callback_handler);
```

Comment:
Github won't let me write this as a suggestion since it involves a deleted line, but I think this would better reproduce the bug:

```
        ThreadSafeReference realm_ref;
        bool called{false};
        auto task = Realm::get_synchronized_realm(config);
        task->start(util::EventLoopDispatcher([&](ThreadSafeReference&& ref, std::exception_ptr error) {
            REQUIRE(!error);
            called = true;
            realm_ref = std::move(ref);
        });
        util::EventLoop::main().run_until([&] {
            return called;
        });
        REQUIRE(called);
        REQUIRE(realm_ref);
        REQUIRE(Realm::get_shared_realm(std::move(realm_ref))->read_group().get_table("class_object"));
```

We specifically want to use a temporary EventLoopDispatcher to ensure that the test code isn't keeping the Scheduler alive. We also don't want to have the mutex or atomic any more as all of the test code should now be running on the main thread, and we want tsan to complain if it isn't.

Reply:
I actually tried something like this, but the timing is so that the bug is not exposed.